nat-lab/tests/utils/connection_tracker.py

```diff
@@ -112,37 +156,54 @@ async def on_stdout(self, stdout: str) -> None:
                         self._sync_event.set()
                         continue
 
+                # skip if we are only interested in new events
+                if (
+                    not self._process_update_events
+                    and event.event_type != EventType.NEW
+                ):
+                    continue
+
                 matching_configs = [
                     cfg for cfg in self._config if cfg.target.partial_eq(connection)
                 ]
                 if not matching_configs:
                     continue
 
-                self._events.append(connection)
+                self._events.append(event)
+                self._check_tcp_state_for_events(event)
 
     async def execute(self) -> None:
         if platform.system() == "Darwin":
             return None
         if not self._config:
-            return
+            return None
 
         await self._process.execute(stdout_callback=self.on_stdout)
 
+    def notify_on_tcp_state(self, state: TcpState, event: asyncio.Event) -> None:
+        """Register an Event to be notified when a specific TCP state is reported"""
+        if state not in self._tcp_state_events:
+            self._tcp_state_events[state] = [event]
+        else:
+            self._tcp_state_events[state].append(event)
```

Comment:
Nit: This is quite common pattern that you can solve by using defaultdict.
```
self._tcp_state_events = defaultdict(list) #instead of {}
... later you can just do this:
self._tcp_state_events[state].append(event)
without worying about creating the key
```